consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -121,48 +124,72 @@ func (x *XDPoS_v2) SignHash(header *types.Header) (hash common.Hash) {
 	return sigHash(header)
 }
 
-func (x *XDPoS_v2) Initial(chain consensus.ChainReader, masternodes []common.Address) error {
+func (x *XDPoS_v2) Initial(chain consensus.ChainReader, header *types.Header) error {
 	log.Info("[Initial] initial v2 related parameters")
 
 	if x.highestQuorumCert.ProposedBlockInfo.Hash != (common.Hash{}) { // already initialized
 		log.Error("[Initial] Already initialized", "x.highestQuorumCert.ProposedBlockInfo.Hash", x.highestQuorumCert.ProposedBlockInfo.Hash)
 		return nil
 	}
 
-	x.lock.Lock()
-	defer x.lock.Unlock()
-	// Check header if it is the first consensus v2 block, if so, assign initial values to current round and highestQC
-
-	log.Info("[Initial] highest QC for consensus v2 first block")
-	// Generate new parent blockInfo and put it into QC
-	// TODO: XIN-147 to initilise V2 engine in a more dynamic way
-	firstV2BlockHeader := chain.GetHeaderByNumber(x.config.V2.SwitchBlock.Uint64())
-	blockInfo := &utils.BlockInfo{
-		Hash:   firstV2BlockHeader.Hash(),
-		Round:  utils.Round(0),
-		Number: firstV2BlockHeader.Number,
-	}
-	quorumCert := &utils.QuorumCert{
-		ProposedBlockInfo: blockInfo,
-		Signatures:        nil,
+	if header.Number.Int64() == x.config.V2.SwitchBlock.Int64() {
+		log.Info("[Initial] highest QC for consensus v2 first block")
+		firstV2BlockHeader := chain.GetHeaderByNumber(header.Number.Uint64())
+		blockInfo := &utils.BlockInfo{
+			Hash:   firstV2BlockHeader.Hash(),
+			Round:  utils.Round(0),
+			Number: firstV2BlockHeader.Number,
+		}
+		quorumCert := &utils.QuorumCert{
+			ProposedBlockInfo: blockInfo,
+			Signatures:        nil,
+		}
+		x.currentRound = 1
+		x.highestQuorumCert = quorumCert
+	} else {
+		log.Info("[Initial] highest QC from current header")
+		quorumCert, round, _, err := x.getExtraFields(header)
+		if err != nil {
+			return err
+		}
+		x.currentRound = round
+		x.highestQuorumCert = quorumCert
 	}
-	x.currentRound = 1
-	x.highestQuorumCert = quorumCert
+
+	//TODO: to be discussed
+
+	checkpointBlockNumber := header.Number.Uint64() - header.Number.Uint64()%x.config.Epoch
```

Comment:
According to PR, we don't need `checkpointBlockNumber = v2.switch` and the correspodning snapshot, do we?
Maybe we need to discuss what if header is very far from v2.switch (e.g. 1000000 is very far from 900). If so the snapshots between 900 and 1000000 are missing.

Reply:
This is initial the closest snapshot. for this case, it's 999450

Reply:
Then snapshot between 900 and 999449 are missing?

Reply:
Sorry maybe this problem is not important.
Here is what I concern:
if V2 Switch = 900, then we want to init gap=450 and its snapshot.
But if a node starts very late, if gets a block 1000000 to init, it should still init gap=450 and its snapshot.
Am I correct?

Reply:
that's correct! other snapshot will be created during sync or it's already in DB

Reply:
however the current implementation, if it fails to load snapshot, it will restore the last snapshot like this 999450 if 1000000 block is the initial block

Reply:
what case it fails to load snapshot 999450? we already have UpdateM1() so snapshot should be there.

Reply:
I understand about snapshot 450. In this PR, we don't need any snapshot <V2.Switch, correct? So I don't need to concern about 450 snapshot.

Reply:
that's discuss in the meeting

Reply:
what case it fails to load snapshot 999450? Usually it shouldn't happen.. So just put as error handling for snapshots over 450. Also we can't call UpdateM1, because it requires to get the gap block state, not current smart contract state. So just use Header's master as work around. It's fine as header's masternodes are finalized, removed all the penalties.

Reply:
> we don't need any snapshot <V2.Switch, correct?

We need snapshot 450, since timeout.GapNumber could = 450.